runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1153,40 +1153,42 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
-                  // type and the value that's being assigned is or might be null, both a run-time
-                  // NULLCHK of the value is required (guarded by a check of whether the
-                  // component type is a value type) and an ArrayStoreCHK are required;
-                  // otherwise, only the ArrayStoreCHK is required.
+                  // Determine whether the value is being copied from the same array that is the target
+                  // of the array element store.  If so, there's no need for an ArrayStoreCHK.
                   //
-                  bool mustFail = false;
-                  if (isArrayStoreCheckNeeded(arrayRefNode, storeValueNode, mustFail, storeClassForCheck, componentClassForCheck))
-                     {
-                     flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
-                     }
+                  TR::Node *storeValueBaseNode = getStoreValueBaseNode(storeValueNode, symRefTab);
 
-                  if ((isCompTypePrimVT != TR_no) && (storeValueConstraint == NULL || !storeValueConstraint->isNonNullObject()))
+                  // If the value was loaded from the same array that is the target of the store, no need
+                  // to perform ArrayStoreCHK.
+                  //
+                  if (storeValueBaseNode == NULL || getValueNumber(storeValueBaseNode) != getValueNumber(arrayRefNode))
                      {
-                     flagsForTransform.set(ValueTypesHelperCallTransform::RequiresNullValueCheck);
+                     // If storing to an array whose component type is or might be a primitive value
+                     // type and the value that's being assigned is or might be null, both a run-time
+                     // NULLCHK of the value is required (guarded by a check of whether the
+                     // component type is a value type) and an ArrayStoreCHK are required;
+                     // otherwise, only the ArrayStoreCHK is required.
+                     //
+                     bool mustFail = false;
+                     if (isArrayStoreCheckNeeded(arrayRefNode, storeValueNode, mustFail, storeClassForCheck, componentClassForCheck))
+                        {
+                        flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
+                        }
                      }
                   }
-               }
 
-            if (isStoreFlattenableArrayElement)
-               {
+                //TODO: Require the <nonNullableArrayNullStoreCheck> non-helper if !canSkipNonNullableArrayNullValueChecks(...)
+                if (canTransformUnflattenedArrayElementLoadStore &&
+                   (isCompTypePrimVT != TR_no) &&
+                   (storeValueConstraint == NULL || !storeValueConstraint->isNonNullObject()))
+                  {
+                  flagsForTransform.set(ValueTypesHelperCallTransform::RequiresNullValueCheck);
+                  }
```

Comment:
The way this was organized before, Value Propagation was able to avoid the call to `<nonNullableArrayNullStoreCheck>` if the value being stored to the array had been loaded from the same array.  That optimization should still be possible, even if the concept of `owningMethodDoesNotContainNonNullableArrayNullValueChecks` is separated from `owningMethodDoesNotContainStoreChecks`.

Reply:
Addressed in 156b344